File: src/test/java/org/rooftop/netx/javasupports/OrchestratorConfigurer.java
package org.rooftop.netx.javasupports;

import org.rooftop.netx.api.Orchestrator;
import org.rooftop.netx.api.OrchestratorFactory;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;

@Configuration
public class OrchestratorConfigurer {

    @Bean
    public Orchestrator<Integer, Integer> intOrchestrator() {
        return OrchestratorFactory.Instance.instance().<Integer>create("intOrchestrator")
            .start(
                request -> request + 1,
                request -> request - 1
            )
            .join(
                request -> request + 1,
                request -> request - 1
            )
            .commit(
                request -> request + 1,
                request -> request - 1
            );
    }
}

File: src/test/kotlin/org/rooftop/netx/client/OrchestratorConfigurer.kt
package org.rooftop.netx.client

import org.rooftop.netx.api.Orchestrate
import org.rooftop.netx.api.Orchestrator
import org.rooftop.netx.api.OrchestratorFactory
import org.springframework.context.annotation.Bean
import reactor.core.publisher.Mono

internal class OrchestratorConfigurer {

    @Bean
    fun sum3Orchestrator(): Orchestrator<Int, Int> {
        return OrchestratorFactory.instance().create<Int>("sum3Orchestrator")
            .startReactive(MonoIntOrchestrator, rollback = { Mono.fromCallable { it - 1 } })
            .joinReactive(MonoIntOrchestrator, rollback = { Mono.fromCallable { it - 1 } })
            .commitReactiveWithContext({ _, request ->
                Mono.fromCallable {
                    request + 1
                }
            })
    }

    object IntOrchestrator : Orchestrate<Int, Int> {

        override fun orchestrate(request: Int): Int = request + 1
    }

    object MonoIntOrchestrator : Orchestrate<Int, Mono<Int>> {

        override fun orchestrate(request: Int): Mono<Int> = Mono.fromCallable { request + 1 }
    }
}
